readline.c

```diff
@@ -30,19 +36,22 @@ readline_next(readline_t * rl) {
   size_t len;
   size_t buffer_len = strlen(rl->buffer);
 
+  assert(rl->buffer != NULL);
+
   while (
-    rl->buffer[cur++] != '\n' && 
+    rl->buffer[cur++] != '\n' &&
     cur <= buffer_len);
 
   len = cur - rl->cursor - 1;
   ret = (char*) malloc(len);
 
-  if (ret == NULL || (len == 0 && cur > buffer_len)) {
+  if (ret == NULL) {
+    return NULL;
+  } else if (len == 0 && cur > buffer_len) {
     free(ret);
     return NULL;
   }
 
-  memset(ret, 0, len);
```

Comment:
memcpy is already overwriting all the buffer. I don't see why is necessary to fill it with 0's when you are going to overwrite it.